File: 162-find-peak-element/162.py
class Solution:
    def bsearch(self, num):
        head = 0
        tail = len(num) - 1

        while head < tail:
            mid = head + (tail - head) // 2
            if num[mid-1] < num[mid] > num[mid+1]:
                return mid
            if num[mid-1] < num[mid] < num[mid+1]:
                head = mid + 1
            elif num[mid-1] > num[mid] > num[mid+1]:
                tail = mid - 1
            else:
                head = mid + 1

        return head

    def findPeakElement(self, nums: List[int]) -> int:
        return self.bsearch(nums)

File: 162-find-peak-element/162.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        if (nums.size() == 1)
            return 0;

        int left = -1;
        int right = nums.size() - 1;
        while (right - left > 1) {
            int lmid = (left + right + 1) / 2;
            int rmid = lmid + 1;
            if (nums[lmid] <= nums[rmid])
                left = lmid;
            else
                right = rmid - 1;
        }

        return right;
    }
};
